Add logging to M.E.Caching.StackExchangeRedis

martincostello opened the pull request:
# Add logging to M.E.Caching.StackExchangeRedis

Support logging in the `RedisCache` class.

## Description

- Add logging to M.E.Caching.StackExchangeRedis.
- Log warning if Redis server version cannot be detected.

Fixes #41723

martincostello committed:
Add logging to M.E.Caching.StackExchangeRedis

- Add logging to M.E.Caching.StackExchangeRedis.
- Log warning if Redis server version cannot be detected.

Relates to #41723.

martincostello committed:
Make logging for RedisCache internal

- Revert changes to public API surface for RedisCache.
- Create new RedisCacheImpl to provide the concrete type for DI.

See https://github.com/dotnet/aspnetcore/issues/41723#issuecomment-1133446437.

BrennanConroy commented on src/Caching/StackExchangeRedis/src/RedisCache.Log.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    private static partial class Log
+    {
+        [LoggerMessage(1, LogLevel.Warning, "Could not determine the Redis server version as the GetServer() operation is not supported.", EventName = "GetServerNotSupported")]
```
I think the log should be less about `GetServer` failing and more about we are falling back to the deprecated `HMSET` command. Or maybe we need two logs.

martincostello replied:
I've changed the error message to mention the different commands. I'm not tied to the specific wording, so happy to change as you deem fit.

halter73 commented on src/Caching/StackExchangeRedis/src/RedisCacheImpl.cs:
```diff
@@ -0,0 +1,15 @@
+
+internal sealed class RedisCacheImpl : RedisCache
+{
+    public RedisCacheImpl(IOptions<RedisCacheOptions> optionsAccessor, ILogger<RedisCacheImpl> logger)
```
This ends up being way cleaner than how we did it for `KestrelServerImpl` though. Leaving the implementation in `RedisCache` and just adding the public constructor here is nice.

I think we should add a `public RedisCacheImpl(IOptions<RedisCacheOptions> optionsAccessor)` constructor just in case someone out there was calling this on an `IServiceCollection` without logging. I think this would be cleaner than `.AddLogging()` in `.AddStackExchangeRedisCache().` If someone already has logging configured, it will light up. Otherwise, it won't add any unnecessary services.

halter73 commented on src/Caching/StackExchangeRedis/src/StackExchangeRedisCacheServiceCollectionExtensions.cs:
```diff
@@ -32,8 +34,16 @@ public static IServiceCollection AddStackExchangeRedisCache(this IServiceCollect
+
         services.Configure(setupAction);
-        services.Add(ServiceDescriptor.Singleton<IDistributedCache, RedisCache>());
+        services.Add(ServiceDescriptor.Singleton<IDistributedCache, RedisCacheImpl>());
```
Is switching the service descriptor to a factory really any better than switching the implementation type?

The runtime type of the object the factory returns would still be `RedisCacheImpl`. And `ImplementationFactory` is declared as just a `Func<IServiceProvider, object>` at compile time. The runtime type of the func is different because of covariance, but who's looking at the runtime type of `ImplementationFactory`? Not the DI system itself. I'd stick with the type over the factory because the DI can optimize it better.

I don't remember anyone complaining about the `KestrelServer` and `KestrelServerImpl` split, and that was even more breaking thinking back. If someone was calling `(KestrelServer)app.Services.GetRequiredService<IServer>()`, it would have broke. At least `(RedisCache)app.Services.GetRequiredService<IDistributedCache>()` still works. I don't think we even made a breaking change announcement for the Kestrel change.

BrennanConroy replied:
The breaking part is doing `app.Services.GetRequiredService<RedisCache>()`, which is not a common pattern probably, but still breaking.

halter73 replied:
> `app.Services.GetRequiredService<RedisCache>()`

I don't think this ever worked if you just called `builder.Services.AddStackExchangeRedisCache()`. If you called, `builder.Services.AddSingleton<RedisCache>()`, it would work. But that still works after this change. Now the only downside is that you don't get logging that way.

martincostello replied:
I just tested this locally, and using `services.Add(ServiceDescriptor.Singleton<IDistributedCache, RedisCache>())` and then doing the below fails:

```csharp
[Fact]
public void AddStackExchangeRedisCache_RegistersRedisCache()
{
    // Arrange
    var services = new ServiceCollection();

    // Act
    services.AddStackExchangeRedisCache(options => { });

    // Assert
    var serviceProvider = services.BuildServiceProvider();
    var redisCache = serviceProvider.GetRequiredService<RedisCache>();

    Assert.NotNull(redisCache);
}
```

``` Microsoft.Extensions.Caching.StackExchangeRedis.CacheServiceExtensionsTests.AddStackExchangeRedisCache_RegistersRedisCache
   Source: CacheServiceExtensionsTests.cs line 17
   Duration: 11 ms

  Message: 
System.InvalidOperationException : No service for type 'Microsoft.Extensions.Caching.StackExchangeRedis.RedisCache' has been registered.

  Stack Trace: 
ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
CacheServiceExtensionsTests.AddStackExchangeRedisCache_RegistersRedisCache() line 27
```

halter73 commented on src/Caching/StackExchangeRedis/src/StackExchangeRedisCacheServiceCollectionExtensions.cs:
```diff
@@ -32,8 +34,16 @@ public static IServiceCollection AddStackExchangeRedisCache(this IServiceCollect
         services.AddOptions();
+
+#if NET7_0_OR_GREATER
+        services.AddLogging();
```
I still don't think we should be adding logging at all in this extension method. I think just lighting up if it's there is fine. But out of curiosity, why the difference between .NET 7 and everything else?

martincostello replied:
> But out of curiosity, why the difference between .NET 7 and everything else?

`AddLogging()` isn't available for `netstandard2.0` and `net472`.

martincostello committed:
Address review feedback

- Update log message.
- Add additional non-logging constructor to RedisCacheImpl.
- Change logger category to RedisCache.
- Remove logging service registrations.
- Add explicit tests for service provider resolution of IDistributedCache.

halter73 reviewed (approved):
Thanks!

BrennanConroy commented:
Thanks @martincostello 😄
